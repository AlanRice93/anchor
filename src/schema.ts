type AttributeCrud = {
  ops?: {
    /**
     * Defaults to optional
     **/
    create?: "required" | "optional" | "omit"
    // Default required
    read?: "required" | "optional" | "omit"
    // Default optional
    update?: "required" | "optional" | "omit"
  }
}

type AttributeNullable = {
  nullable?: boolean
}

export type AttributeDefinition = AttributeNullable &
  AttributeCrud &
  (
    | AttributeStringDefinition
    | AttributeObjectDefinition
    | AttributeArrayDefinition
    | AttributeBooleanDefinition
    | AttributeIntegerDefinition
    | AttributeNumberDefinition
  )

export type AttributeStringDefinition = {
  type: "string"
  minLength?: number
  maxLength?: number
  pattern?: string
  format?:
    | "uri"
    | "email"
    | "date-time"
    | "time"
    | "date"
    | "duration"
    | "hostname"
    | "uuid"
    | "uri"
    | "uri-reference"
    | "ipv4"
    | "ipv6"
    | "regex"
}

type AttributeNumberDefinitionOpts = {
  minimum?: number
  exclusiveMinimum?: number
  maximum?: number
  exclusiveMaximum?: number
  multipleOf?: number
}

export type AttributeNumberDefinition = AttributeNumberDefinitionOpts & {
  type: "number"
}

export type AttributeIntegerDefinition = AttributeNumberDefinitionOpts & {
  type: "integer"
}

export type AttributeBooleanDefinition = {
  type: "boolean"
}

export type AttributeObjectDefinition = {
  type: "object"
  properties: Record<string, AttributeDefinition>
}

export type AttributeArrayDefinition = {
  type: "array"
  items: AttributeDefinition
}

export type RelationshipDefinition = {
  type: "many" | "one"
  resource: string
  inverse?: string
  // Default false
  nullable?: boolean
}

export type ResourceDefinition = {
  type: string
  ops?: {
    /**
     * Specifies whether this resource can be created
     **/
    create?: boolean
    /**
     * Specifies whether this resource can be read
     **/
    read?: boolean
    /**
     * Specifies whether this resource can be updated
     **/
    update?: boolean
    /**
     * Specifies whether this resource can be deleted
     **/
    delete?: boolean
  }
  /**
   * Attributes that this resource supports
   **/
  attributes: Record<string, AttributeDefinition>
  relationships: Record<string, RelationshipDefinition>
}

export type SchemaDefinition = ResourceDefinition[]
