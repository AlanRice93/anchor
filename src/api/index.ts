export * from "./operation"
export * from "./handler"
export * from "./validate"
export * as Errors from "./error"
