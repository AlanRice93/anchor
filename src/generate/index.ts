export { generate as typescript } from "./typescript"
export { generate as orbit } from "./orbit"
export { generate as jsonschema } from "./jsonschema"
