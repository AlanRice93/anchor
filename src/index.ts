export * from "./schema"
export * as Generate from "./generate"
export * as Api from "./api"
export * as Config from "./config"
