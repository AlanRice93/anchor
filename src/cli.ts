#!/usr/bin/env node

import { program } from "commander"
import { writeFileSync } from "fs"
import path from "path"
import { Generate } from "./generate/generate"
import * as Anchor from "./index"
import chalk from "chalk"

program.command("gen <src>").action((src: string) => {
  const config = Anchor.Config.load(path.resolve(src))
  console.log(
    chalk.cyan(
      "Resources:",
      config.resources.map((item) => item.type).join(", ")
    )
  )
  for (let [type, outputs] of Object.entries(config.outputs)) {
    console.log(chalk.green("Generating", type))
    const g: Generate = Anchor.Generate[type]
    const result = g(config.resources)
    for (let output of outputs) {
      writeFileSync(output, result)
      console.log(chalk.gray("-> Wrote", output))
    }
  }
})

program.parse(process.argv)
