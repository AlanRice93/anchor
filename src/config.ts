import { rmSync } from "fs"
import { Project } from "ts-morph"
import { SchemaDefinition } from "./schema"
import * as Generate from "./generate"

export type Definition = {
  resources: SchemaDefinition
  outputs: { [key in keyof typeof Generate]?: string[] }
}

export function define(config: Definition) {
  return config
}

export function load(path: string): Definition {
  const project = new Project({})
  const source = project.addSourceFileAtPath(path)
  source.emitSync()
  const [compiled] = source.getEmitOutput().getOutputFiles()
  const config = require(compiled.getFilePath())
  rmSync(compiled.getFilePath())
  return config.default
}
