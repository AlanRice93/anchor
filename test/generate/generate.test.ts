import { Generate, SchemaDefinition } from "../../src"

const schema: SchemaDefinition = [
  {
    type: "user",
    attributes: {
      name: {
        type: "string",
      },
    },
    relationships: {
      posts: {
        type: "many",
        resource: "article",
        inverse: "author",
      },
    },
  },
  {
    type: "post",
    attributes: {
      title: {
        type: "string",
      },
    },
    relationships: {
      author: {
        type: "one",
        resource: "user",
        inverse: "articles",
      },
    },
  },
]

it("typescript", () => {
  const output = Generate.typescript(schema)
  expect(output).toMatchSnapshot()
})

it("jsonschema", () => {
  const output = Generate.jsonschema(schema)
  expect(output).toMatchSnapshot()
})

it("orbit", () => {
  const output = Generate.orbit(schema)
  expect(output).toMatchSnapshot()
})
