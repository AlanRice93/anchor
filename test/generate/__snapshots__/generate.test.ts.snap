// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jsonschema 1`] = `
"{
  \\"UserDataCreate\\": {
    \\"title\\": \\"UserDataCreate\\",
    \\"additionalProperties\\": false,
    \\"type\\": \\"object\\",
    \\"properties\\": {
      \\"id\\": {
        \\"type\\": \\"string\\",
        \\"format\\": \\"uuid\\"
      },
      \\"type\\": {
        \\"const\\": \\"user\\"
      },
      \\"attributes\\": {
        \\"type\\": \\"object\\",
        \\"additionalProperties\\": false,
        \\"properties\\": {
          \\"name\\": {
            \\"type\\": \\"string\\"
          }
        },
        \\"required\\": []
      },
      \\"relationships\\": {
        \\"type\\": \\"object\\",
        \\"additionalProperties\\": false,
        \\"properties\\": {
          \\"posts\\": {
            \\"type\\": \\"object\\",
            \\"additionalProperties\\": false,
            \\"properties\\": {
              \\"data\\": {
                \\"type\\": \\"object\\",
                \\"additionalProperties\\": false,
                \\"required\\": [
                  \\"id\\",
                  \\"type\\"
                ],
                \\"properties\\": {
                  \\"id\\": {
                    \\"type\\": \\"string\\"
                  },
                  \\"type\\": {
                    \\"const\\": \\"article\\"
                  }
                }
              }
            }
          }
        }
      }
    },
    \\"required\\": [
      \\"id\\",
      \\"attributes\\",
      \\"relationships\\",
      \\"type\\"
    ]
  },
  \\"PostDataCreate\\": {
    \\"title\\": \\"PostDataCreate\\",
    \\"additionalProperties\\": false,
    \\"type\\": \\"object\\",
    \\"properties\\": {
      \\"id\\": {
        \\"type\\": \\"string\\",
        \\"format\\": \\"uuid\\"
      },
      \\"type\\": {
        \\"const\\": \\"post\\"
      },
      \\"attributes\\": {
        \\"type\\": \\"object\\",
        \\"additionalProperties\\": false,
        \\"properties\\": {
          \\"title\\": {
            \\"type\\": \\"string\\"
          }
        },
        \\"required\\": []
      },
      \\"relationships\\": {
        \\"type\\": \\"object\\",
        \\"additionalProperties\\": false,
        \\"properties\\": {
          \\"author\\": {
            \\"type\\": \\"object\\",
            \\"additionalProperties\\": false,
            \\"properties\\": {
              \\"data\\": {
                \\"type\\": \\"object\\",
                \\"additionalProperties\\": false,
                \\"required\\": [
                  \\"id\\",
                  \\"type\\"
                ],
                \\"properties\\": {
                  \\"id\\": {
                    \\"type\\": \\"string\\"
                  },
                  \\"type\\": {
                    \\"const\\": \\"user\\"
                  }
                }
              }
            }
          }
        }
      }
    },
    \\"required\\": [
      \\"id\\",
      \\"attributes\\",
      \\"relationships\\",
      \\"type\\"
    ]
  }
}"
`;

exports[`orbit 1`] = `
"import { ModelDefinition } from \\"@orbit/records\\";
export const Models: Record<string, ModelDefinition> = {
  user: {
    attributes: {
      name: {
        type: \\"string\\",
      },
    },
    relationships: {
      posts: {
        inverse: \\"author\\",
        model: \\"article\\",
        kind: \\"hasMany\\",
      },
    },
  },
  post: {
    attributes: {
      title: {
        type: \\"string\\",
      },
    },
    relationships: {
      author: {
        inverse: \\"articles\\",
        model: \\"user\\",
        kind: \\"hasOne\\",
      },
    },
  },
};
"
`;

exports[`typescript 1`] = `
"export type UserDataRead = {
  id: string;
  type: \\"user\\";
  attributes: {
    name: string;
  };
  relationships: {
    posts?: {
      data: {
        id: string;
        type: \\"article\\";
      }[];
    };
  };
};
export type UserDataCreate = {
  id: string;
  type: \\"user\\";
  attributes: {
    name?: string;
  };
  relationships: {
    posts?: {
      data: {
        id: string;
        type: \\"article\\";
      }[];
    };
  };
};
export type UserDataUpdate = {
  id: string;
  type: \\"user\\";
  attributes: {
    name?: string;
  };
  relationships: {
    posts?: {
      data: {
        id: string;
        type: \\"article\\";
      }[];
    };
  };
};
export type PostDataRead = {
  id: string;
  type: \\"post\\";
  attributes: {
    title: string;
  };
  relationships: {
    author?: {
      data: {
        id: string;
        type: \\"user\\";
      };
    };
  };
};
export type PostDataCreate = {
  id: string;
  type: \\"post\\";
  attributes: {
    title?: string;
  };
  relationships: {
    author?: {
      data: {
        id: string;
        type: \\"user\\";
      };
    };
  };
};
export type PostDataUpdate = {
  id: string;
  type: \\"post\\";
  attributes: {
    title?: string;
  };
  relationships: {
    author?: {
      data: {
        id: string;
        type: \\"user\\";
      };
    };
  };
};
"
`;
